### PR TITLE
Fix: fix incorrect conditional logic for dynamic blocks iterating on `var.ec2_tag_set` and `var.ec2_tag_filter`

### PR DESCRIPTION
**Pull request by StephenTan-TW**
## what

Original PR description:

> In order to ensure that tagging is processed correctly, I have created a PR of working code. 
> For some reason, we need to disable conditionals for ec2_tag_filter variable and we need to set the following conditional for ec2_tag_set: ```var.ec2_tag_set == null ? [] : [var.ec2_tag_set]```
> Note that I have had to do a huge rebase when pulling the upstream code. No idea what is going on.

- Swap second and third arguments in ternary operator for `var.ec2_tag_set` and `var.ec2_tag_filter`, which are incorrectly supplying an empty list if their lengths are > 0.
- Use `lookup` function in dynamic blocks in order to introduce more robustness.

## why

* If `var.ec2_tag_set` and `var.ec2_tag_filter` are supplied as non-empty lists, the `ec2_tag_filter` blocks will not be populated, since the ternary operator will evaluate to empty lists, and hence the dynamic block will produce no such blocks.

### CODE REVIEW

**Comment by StephenTan-TW** [comment]
these 34 commits clearly should not be going into your history. I am trying to figure out what to do. Should I just do a huge rebase to squash everything out? Not sure what do do here.

File: main.tf [hunk 1]
```diff
@@ -178,7 +178,8 @@ resource "aws_codedeploy_deployment_group" "default" {
   # Note that you cannot have both ec_tag_filter and ec2_tag_set vars set!
   # See https://docs.aws.amazon.com/cli/latest/reference/deploy/create-deployment-group.html for details
   dynamic "ec2_tag_filter" {
-    for_each = length(var.ec2_tag_filter) > 0 ? [] : var.ec2_tag_filter
+    for_each = var.ec2_tag_filter
+
     content {
       key   = ec2_tag_filter.value["key"]
       type  = ec2_tag_filter.value["type"]
```
Comment (nitrocode) [inline main.tf:185]: couldn't we use lookups here that result in null if the key is missing?

```suggestion
      type  = lookup(ec2_tag_filter.value, "type", null)
```

File: main.tf [hunk 2]
```diff
@@ -178,7 +178,8 @@ resource "aws_codedeploy_deployment_group" "default" {
   # Note that you cannot have both ec_tag_filter and ec2_tag_set vars set!
   # See https://docs.aws.amazon.com/cli/latest/reference/deploy/create-deployment-group.html for details
   dynamic "ec2_tag_filter" {
-    for_each = length(var.ec2_tag_filter) > 0 ? [] : var.ec2_tag_filter
```
Comment (nitrocode) [inline main.tf:181]: if the length is greater than 0, then it should use var.ec2_tag_filter. in the deleted logic, it was not used if it was greater than zero.

File: main.tf [hunk 3]
```diff
@@ -189,7 +190,7 @@ resource "aws_codedeploy_deployment_group" "default" {
   # Note that you cannot have both ec_tag_filter and ec2_tag_set vars set!
   # See https://docs.aws.amazon.com/cli/latest/reference/deploy/create-deployment-group.html for details
   dynamic "ec2_tag_set" {
-    for_each = length(var.ec2_tag_set) > 0 ? [] : var.ec2_tag_set
+    for_each = var.ec2_tag_set == null ? [] : [var.ec2_tag_set]
```
Comment (nitrocode) [inline main.tf:193]:
```suggestion
    for_each = length(var.ec2_tag_set) > 0 ? [var.ec2_tag_set] : []
```

File: main.tf [hunk 4]
```diff
@@ -178,7 +178,8 @@ resource "aws_codedeploy_deployment_group" "default" {
   # Note that you cannot have both ec_tag_filter and ec2_tag_set vars set!
   # See https://docs.aws.amazon.com/cli/latest/reference/deploy/create-deployment-group.html for details
   dynamic "ec2_tag_filter" {
-    for_each = length(var.ec2_tag_filter) > 0 ? [] : var.ec2_tag_filter
+    for_each = var.ec2_tag_filter
```
Comment (nitrocode) [inline main.tf:181]:
```suggestion
    for_each = length(var.ec2_tag_filter) > 0 ? [var.ec2_tag_filter] : []
```

File: main.tf (hunk 2 above)
Comment (StephenTan-TW) [reply main.tf:181]: Oh I see. Good grief. Literally the opposite of what we wanted. I feel a bit dumb now

File: main.tf (hunk 1 above)
Comment (StephenTan-TW) [reply main.tf:185]: ah... yes, sure. I'll amend the PR in a moment. Thank you for this! This code is going to be pretty robust once we're done! :)

**Comment by StephenTan-TW** [comment]
Hm.... for the lookups - I'm now getting: 

```On .terraform/modules/code-deploy-stargate/main.tf line 184: An argument or block definition is required here. To set an argument, use the equals sign "=" to introduce the argument value.```

File: main.tf [hunk 5]
```diff
@@ -178,26 +178,27 @@ resource "aws_codedeploy_deployment_group" "default" {
...
-    for_each = length(var.ec2_tag_filter) > 0 ? [] : var.ec2_tag_filter
+    for_each = length(var.ec2_tag_filter) > 0 ? var.ec2_tag_filter : []
+
     content {
-      key   = ec2_tag_filter.value["key"]
-      type  = ec2_tag_filter.value["type"]
-      value = ec2_tag_filter.value["value"]
+      lookup(ec2_tag_filter["key"], "key", null)
+      lookup(ec2_tag_filter["type"], "type", null)
+      lookup(ec2_tag_filter["value"], "value", null)
```
Comment (korenyoni) [inline main.tf:186]:
```suggestion
      key = lookup(ec2_tag_filter["key"], "key", null)
      type = lookup(ec2_tag_filter["type"], "type", null)
      value = lookup(ec2_tag_filter["value"], "value", null)
```

File: main.tf [hunk 6]
```diff
@@ -178,26 +178,27 @@ resource "aws_codedeploy_deployment_group" "default" {
...
     content {
       dynamic "ec2_tag_filter" {
         for_each = ec2_tag_set.value.ec2_tag_filter
         content {
-          key   = ec2_tag_filter.value["key"]
-          type  = ec2_tag_filter.value["type"]
-          value = ec2_tag_filter.value["value"]
+          lookup(ec2_tag_set["key"], "key", null)
+          lookup(ec2_tag_set["type"], "type", null)
+          lookup(ec2_tag_set["value"], "value", null)
```
Comment (korenyoni) [inline main.tf:201]:
```suggestion
          key = lookup(ec2_tag_set["key"], "key", null)
          type = lookup(ec2_tag_set["type"], "type", null)
          value = lookup(ec2_tag_set["value"], "value", null)
```

**Review by korenyoni — CHANGES_REQUESTED** [review]
Minor changes needed to get this working

**Comment by korenyoni** [comment]
Thanks 👍 

Let's see if tests pass now

File: main.tf [hunk 7]
```diff
@@ -178,26 +178,27 @@ resource "aws_codedeploy_deployment_group" "default" {
...
     content {
       dynamic "ec2_tag_filter" {
         for_each = ec2_tag_set.value.ec2_tag_filter
         content {
-          key   = ec2_tag_filter.value["key"]
-          type  = ec2_tag_filter.value["type"]
-          value = ec2_tag_filter.value["value"]
+          key = lookup(ec2_tag_set.value, "key", null)
+          type = lookup(ec2_tag_set.value, "type", null)
+          value = lookup(ec2_tag_set."value, "value", null)
```
Comment (korenyoni) [inline main.tf:201]:
```suggestion
          key = lookup(ec2_tag_filter.value, "key", null)
          type = lookup(ec2_tag_filter.value, "type", null)
          value = lookup(ec2_tag_filter."value, "value", null)
```

**Comment by StephenTan-TW** [comment]
ach. Not my day today. Yes, all working now. Thank you

**Review by korenyoni — APPROVED** [review]
LGTM.

**Comment by StephenTan-TW** [comment]
Please for the love of god can we merge and create a release now? Would really like to fix my broken CodeDeploy modules - they are pointing to my personal forked repo tags atm which triggers my OCD. Thank you

**Comment by korenyoni** [comment]
@StephenTan-TW I'm triple checking everything because as simple as this PR is, I walked into this PR without initially understanding the problem.

So, it's better to check things thoroughly rather than create a release where something was missed, where other users could run into a problem. Please be patient and think of the other users of this module.

**Comment by StephenTan-TW** [comment]
I get it - not trying to cause any problems. Just very tired. Appreciate the fresh eyes on this. It was a case of not being able to see the wood for the trees. Thank you again.

**Comment by korenyoni** [comment]
Np, thank you for identifying and fixing this bug!

**Comment by StephenTan-TW** [comment]
you guys do great work. This is Open Source at it's best. Appreciate it and very happy to contribute. Keep it up!

**Pull request merged**